fuse/test/file_lock_test.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/internal/testutil"
+)
+
+func TestFlockExclusive(t *testing.T) {
+	cmd, err := exec.LookPath("flock")
+	if err != nil {
+		t.Skip("flock command not found.")
+	}
+	tc := NewTestCase(t)
+	defer tc.Cleanup()
+
+	contents := []byte{1, 2, 3}
+	tc.WriteFile(tc.origFile, []byte(contents), 0700)
+
+	f, err := os.OpenFile(tc.mountFile, os.O_WRONLY, 0)
+	if err != nil {
+		t.Fatalf("OpenFile(%q): %v", tc.mountFile, err)
+	}
+	defer f.Close()
+
+	if err = syscall.Flock(int(f.Fd()), syscall.LOCK_EX); err != nil {
+		t.Errorf("Flock returned: %v", err)
+		return
+	}
+
+	if out, err := runExternalFlock(cmd, tc.mountFile); !bytes.Contains(out, []byte("failed to get lock")) {
+		t.Errorf("runExternalFlock(%q): %s (%v)", tc.mountFile, out, err)
+	}
+}
+
+func runExternalFlock(flockPath, fname string) ([]byte, error) {
+	f, err := os.OpenFile(fname, os.O_WRONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	cmd := exec.Command(flockPath, "--verbose", "--exclusive", "--nonblock", "3")
+	cmd.Env = append(cmd.Env, "LC_ALL=C") // in case the user's shell language is different
+	cmd.ExtraFiles = []*os.File{f}
+	return cmd.CombinedOutput()
+}
+
+type lockingNode struct {
+	nodefs.Node
+	flockInvoked bool
+}
+
+func (n *lockingNode) Open(flags uint32, context *fuse.Context) (file nodefs.File, code fuse.Status) {
+	return nodefs.NewDataFile([]byte("hello world")), fuse.OK
+}
+
+func (n *lockingNode) GetLk(file nodefs.File, owner uint64, lk *fuse.FileLock, flags uint32, out *fuse.FileLock, context *fuse.Context) (code fuse.Status) {
+	n.flockInvoked = true
+	return fuse.OK
+}
+
+func (n *lockingNode) SetLk(file nodefs.File, owner uint64, lk *fuse.FileLock, flags uint32, context *fuse.Context) (code fuse.Status) {
+	n.flockInvoked = true
+	return fuse.OK
+}
+
+func (n *lockingNode) SetLkw(file nodefs.File, owner uint64, lk *fuse.FileLock, flags uint32, context *fuse.Context) (code fuse.Status) {
+	n.flockInvoked = true
+	return fuse.OK
+}
+
+func TestFlockInvoked(t *testing.T) {
+	flock, err := exec.LookPath("flock")
+	if err != nil {
+		t.Skip("flock command not found.")
+	}
+
+	dir := testutil.TempDir()
+	defer os.RemoveAll(dir)
+
+	opts := &nodefs.Options{
+		Owner: fuse.CurrentOwner(),
+		Debug: testutil.VerboseTest(),
+	}
+
+	root := nodefs.NewDefaultNode()
+	s, _, err := nodefs.MountRoot(dir, root, opts)
+	if err != nil {
+		t.Fatalf("MountRoot: %v", err)
+	}
+	go s.Serve()
+	if err := s.WaitMount(); err != nil {
+		t.Fatal("WaitMount", err)
+	}
+	defer s.Unmount()
+
+	node := &lockingNode{
+		Node:    nodefs.NewDefaultNode(),
+		flockInvoked: false,
+	}
+	root.Inode().NewChild("foo", false, node)
+
+	realPath := filepath.Join(dir, "foo")
+	cmd:=exec.Command(flock, "--nonblock", realPath, "echo", "locked")
+	out, err :=cmd.CombinedOutput()
+	if err!=nil {
+		t.Fatalf("flock %v: %v",err, string(out))
+	}
+	if !node.flockInvoked {
+		t.Fatalf("flock is not invoked")
+	}
```

Comment:
this test makes sure the locking really goes through FUSE (which is good), but I can't see the functionality of GetLk/SetLk/SetLkw tested separately. From the signatures, it seems one can have different locks on ranges of the file, and it would be good to exercise that too. Can you think of a way to add coverage for those features? Maybe you could record the ranges of the locks and verify they are as you expect?

Reply:
We added additional testing to ensure SetLk and SetLkw are called but we couldn't find a way to use flock to invoke GetLk. We tried cgo in an attempt to call fcntl.c but cgo isnt allowed in test files. Maybe you have some other ideas?